lib/events/athena/consumer.go

```diff
@@ -0,0 +1,457 @@
+	return s.eventsChan
+}
+
+// fromSQS receives messages from SQS and sends it on eventsC channel.
+// It runs until context is canceled (via timeout) or when maxItems is reached.
+// MaxItems is soft limit and can happen that it will return more items then MaxItems.
+func (s *sqsMessagesCollector) fromSQS(ctx context.Context) {
+	// Errors should be immediately process by error handling loop, so 10 size
+	// should be enough to not cause blocking.
+	errorsC := make(chan error, 10)
+	defer close(errorsC)
+
+	// errhandle loop for receiving single event errors.
+	go func() {
+		s.errHandlingFn(ctx, errorsC)
+	}()
+	eventsC := s.eventsChan
+
+	count := 0
+	countMu := sync.Mutex{}
+
+	// wokerCtx is mechanism to stop other workers when maxItems is reached.
+	wokerCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	wg := sync.WaitGroup{}
+	wg.Add(maxNumberOfWorkers)
+	for i := 0; i < maxNumberOfWorkers; i++ {
+		go func(i int) {
+			defer wg.Done()
+			for {
+				if wokerCtx.Err() != nil {
+					return
+				}
+
+				// If there is not enough time to process receiveMessage call
+				// we can return immediately. It's added because if
+				// receiveMessages is canceled message is marked as not
+				// processed after VisibilitTimeout (equal to BatchInterval).
+				if deadline, ok := wokerCtx.Deadline(); ok && time.Until(deadline) <= s.config.waitOnReceiveDuration {
+					return
+				}
+				events, err := s.receiveMessages(wokerCtx)
+				if err != nil {
+					// TODO(tobiaszheller): maybe we need to check for other errors than cancel as well
+					// from aws SDK.
+					if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+						return
+					}
+					errorsC <- trace.Wrap(err)
+				}
+				for _, e := range events {
+					if e.err != nil {
+						errorsC <- e.err
+						continue
+					}
+					eventsC <- eventAndAckID{
+						event:         e.data,
+						receiptHandle: e.receiptHandle,
+					}
+				}
+
+				countMu.Lock()
+				count += len(events)
+				if count >= s.config.batchMaxItems {
+					countMu.Unlock()
+					cancel()
+					return
+				}
+				countMu.Unlock()
+			}
+		}(i)
+	}
+	wg.Wait()
+	close(eventsC)
+}
+
+type sqsMessageWithError struct {
+	// receiptHandle is used to delete message from queue
+	receiptHandle string
+	data          *eventParquet
+	err           error
+}
+
+func (s *sqsMessagesCollector) receiveMessages(ctx context.Context) ([]sqsMessageWithError, error) {
+	out := make([]sqsMessageWithError, 0, 10)
+	sqsOut, err := s.config.sqsReceiver.ReceiveMessage(ctx, &sqs.ReceiveMessageInput{
+		QueueUrl:              aws.String(s.config.queueURL),
+		MaxNumberOfMessages:   maxNumberOfMessagesFromReceive,
+		WaitTimeSeconds:       s.config.waitOnReceiveTimeout,
+		VisibilityTimeout:     s.config.visibilityTimeout,
+		MessageAttributeNames: []string{payloadTypeAttr},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if len(sqsOut.Messages) < 1 {
```

Comment:
Why not `len(sqsOut.Messages) == 0`?

Can `len(sqsOut.Messages)` ever be negative?